src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/DataSetService.cs

```diff
@@ -99,6 +107,23 @@ private static DataSetViewModel MapDataSet(DataSet dataSet)
         };
     }
 
+    private async Task<Either<ActionResult, DataSetVersion>> CheckVersionExists(
+        Guid dataSetId,
+        string dataSetVersion)
+    {
+        var dataSetVersions = await _publicDataDbContext.DataSetVersions
+            .Where(dsv => dsv.DataSetId == dataSetId)
+            .Where(ds => ds.Status == DataSetVersionStatus.Published
+                || ds.Status == DataSetVersionStatus.Unpublished
+                || ds.Status == DataSetVersionStatus.Deprecated)
+            .ToListAsync();
+
+        return await dataSetVersions
+            .Where(dsv => dsv.Version == dataSetVersion)
+            .SingleOrDefault()
+            .OrNotFound();
```

Comment:
Not sure this is right. We're bringing back all the data set's versions into memory and checking them there, rather than just directly querying them in the database?

Think this can just be:

```cs
return await _publicDataDbContext.DataSetVersions
    .Where(dsv => dsv.DataSetId == dataSetId)
    .Where(ds => ds.Status == DataSetVersionStatus.Published
        || ds.Status == DataSetVersionStatus.Unpublished
        || ds.Status == DataSetVersionStatus.Deprecated)
    .Where(dsv => dsv.Version == dataSetVersion)
    .SingleOrNotFound();
```

Plus we get to use `SingleOrNotFound` as well

Reply:
Can't do that as it can't convert `dsv.Version == dataSetVersion` into an expression translatable into SWL :/